File: app/src/socle/sql.js
/**
 * Vos propres tables.
 *
 * Votre groupe est PROPRIÉTAIRE d'un schéma PostgreSQL qui porte son code.
 * Vous y créez vos tables, vous les lisez, vous les écrivez. Vous lisez aussi
 * le socle (`public.`) et les schémas des cinq autres groupes.
 *
 * Vous n'avez pas à préfixer vos tables : `offres` désigne déjà la table
 * `offres` de votre schéma.
 *
 * La règle de sécurité : NE JAMAIS concaténer une valeur dans une requête.
 * Les paramètres nommés `:nom` sont là pour ça, et ils gèrent les accents,
 * les apostrophes et les pourcents sans que vous ayez à y penser.
 *
 *   ✅ sql('select * from offres where statut = :s', { s: statut.value })
 *   ❌ sql(`select * from offres where statut = '${statut.value}'`)
 */

import { onMounted, reactive } from 'vue'
import { post } from './api'

/**
 * Exécute une requête et renvoie les lignes.
 *
 *   const offres = await sql('select * from offres order by id desc')
 *   const trouves = await sql(
 *     'select * from candidats where nom ilike :motif',
 *     { motif: `%${recherche.value}%` },
 *   )
 */
export async function sql(requete, params = {}) {
  const reponse = await post('/sql', { query: requete, params })
  return reponse.rows
}

/** Comme `sql`, mais renvoie aussi les colonnes et le nombre de lignes. */
export const sqlComplet = (requete, params = {}) =>
  post('/sql', { query: requete, params })

/** Applique plusieurs instructions d'un coup — votre fichier schema.sql. */
export const sqlScript = (script) => post('/sql/script', { script })

// ── Le CRUD courant, pour ne pas écrire des INSERT à la main ─────────────────

const NOM_VALIDE = /^[a-z_][a-z0-9_]*$/

function verifierNom(table) {
  if (!NOM_VALIDE.test(table)) {
    throw new Error(
      `« ${table} » n'est pas un nom de table valide : minuscules, chiffres et ` +
      `tirets bas uniquement.`,
    )
  }
}

/**
 * insert('candidats', { nom: 'Meyer', prenom: 'Léa' })
 * Renvoie la ligne créée, identifiant compris.
 */
export async function insert(table, objet) {
  verifierNom(table)
  const colonnes = Object.keys(objet)
  if (!colonnes.length) throw new Error('Rien à insérer : l’objet est vide.')

  const lignes = await sql(
    `insert into ${table} (${colonnes.join(', ')}) ` +
    `values (${colonnes.map((c) => `:${c}`).join(', ')}) returning *`,
    objet,
  )
  return lignes[0]
}

/**
 * update('candidatures', 12, { statut: 'entretien' })
 * Renvoie la ligne modifiée.
 */
export async function update(table, id, objet) {
  verifierNom(table)
  const colonnes = Object.keys(objet)
  if (!colonnes.length) throw new Error('Rien à modifier : l’objet est vide.')

  const lignes = await sql(
    `update ${table} set ${colonnes.map((c) => `${c} = :${c}`).join(', ')} ` +
    `where id = :__id returning *`,
    { ...objet, __id: id },
  )
  return lignes[0]
}

/** remove('offres', 3) — supprime une ligne par son identifiant. */
export async function remove(table, id) {
  verifierNom(table)
  await sql(`delete from ${table} where id = :__id`, { __id: id })
  return true
}

// ── La requête réactive, pour les écrans ─────────────────────────────────────

/**
 * Charge une requête et la garde à jour.
 *
 *   const offres = useQuery('select * from offres order by id desc')
 *   // offres.rows · offres.loading · offres.error · offres.refresh()
 *
 * Pour des paramètres qui changent, passez une FONCTION : elle est réévaluée
 * à chaque rafraîchissement.
 *
 *   const offres = useQuery(
 *     'select * from offres where statut = :s',
 *     () => ({ s: statut.value }),
 *   )
 *   watch(statut, offres.refresh)
 */
export function useQuery(requete, params = {}, { chargerAuDemarrage = true } = {}) {
  const etat = reactive({ rows: [], columns: [], loading: false, error: null })

  etat.refresh = async () => {
    etat.loading = true
    etat.error = null
    try {
      const reponse = await sqlComplet(
        requete, typeof params === 'function' ? params() : params,
      )
      etat.rows = reponse.rows
      etat.columns = reponse.columns
    } catch (erreur) {
      etat.error = erreur.message
      etat.rows = []
    } finally {
      etat.loading = false
    }
  }

  if (chargerAuDemarrage) onMounted(etat.refresh)
  return etat
}

File: api/socle/routers/sql.py
"""
L'accès SQL des groupes à leur propre schéma.

Chaque groupe est PROPRIÉTAIRE d'un schéma PostgreSQL portant son code. Il y
crée ses tables, les lit, les écrit. Il lit aussi le socle (`public`) et les
schémas des cinq autres groupes, ce qui permet les jointures inter-modules à
partir du moment où chacun a de la matière.

L'isolation n'est pas assurée par ce fichier : elle est assurée par PostgreSQL.
Chaque requête tourne sous le rôle `grp_<code>` du groupe appelant, qui n'a
aucun droit d'écriture ailleurs que chez lui. Une requête qui tenterait de
modifier le socle est refusée par la base, avec un message explicite.
"""

from fastapi import APIRouter, HTTPException
from sqlalchemy import text
from sqlalchemy.exc import SQLAlchemyError

from .. import schemas
from ..config import LIGNES_MAX_SQL
from ..db import session_groupe
from ..security import GroupeAppelant

router = APIRouter()


def _message_postgres(erreur: SQLAlchemyError) -> str:
    """Le message brut de PostgreSQL : c'est lui qui apprend quelque chose."""
    origine = getattr(erreur, "orig", None)
    return str(origine or erreur).strip()


@router.post("/sql", response_model=schemas.ReponseSql, tags=["SQL des modules"],
             summary="Exécuter une requête sur le schéma de votre module")
def executer(requete: schemas.RequeteSql, groupe: GroupeAppelant) -> schemas.ReponseSql:
    """
    Une seule instruction SQL, avec ses paramètres nommés.

    Les paramètres s'écrivent `:nom` et se passent dans `params`. Ne jamais
    concaténer une valeur saisie par l'utilisateur dans la requête : c'est
    exactement ce que ces paramètres évitent.

        {
          "query": "select * from candidats where statut = :statut order by nom",
          "params": { "statut": "nouveau" }
        }

    La recherche par motif fonctionne de la même façon :

        {
          "query": "select * from candidats where nom ilike :motif",
          "params": { "motif": "%mar%" }
        }

    Une instruction qui ne renvoie pas de lignes (`insert`, `update`, `delete`,
    `create table`) renvoie `rows: []` et le nombre de lignes touchées.
    """
    with session_groupe(groupe) as session:
        try:
            resultat = session.execute(text(requete.query), requete.params or {})
        except SQLAlchemyError as erreur:
            session.rollback()
            raise HTTPException(400, _message_postgres(erreur)) from None

        if resultat.returns_rows:
            colonnes = list(resultat.keys())
            lignes = [dict(zip(colonnes, ligne)) for ligne in resultat.fetchmany(LIGNES_MAX_SQL)]
            tronque = resultat.fetchone() is not None
            session.commit()
            return schemas.ReponseSql(
                columns=colonnes, rows=lignes, row_count=len(lignes), truncated=tronque
            )

        touchees = resultat.rowcount if resultat.rowcount and resultat.rowcount > 0 else 0
        session.commit()
        return schemas.ReponseSql(columns=[], rows=[], row_count=touchees, truncated=False)


@router.post("/sql/script", response_model=schemas.ReponseScriptSql, tags=["SQL des modules"],
             summary="Exécuter plusieurs instructions — pour appliquer votre schema.sql")
def executer_script(script: schemas.ScriptSql, groupe: GroupeAppelant) -> schemas.ReponseScriptSql:
    """
    Plusieurs instructions séparées par des points-virgules, sans paramètres.

    C'est l'endpoint qu'utilise la console SQL du socle pour appliquer le
    fichier `schema.sql` d'un module. Tout passe dans une seule transaction :
    si une instruction échoue, aucune n'est appliquée.
    """
    texte = script.script.strip()
    if not texte:
        raise HTTPException(400, "Le script est vide.")

    with session_groupe(groupe) as session:
        try:
            session.connection().exec_driver_sql(texte)
        except SQLAlchemyError as erreur:
            session.rollback()
            raise HTTPException(400, _message_postgres(erreur)) from None
        session.commit()

    instructions = len([i for i in texte.split(";") if i.strip()])
    return schemas.ReponseScriptSql(
        statements=instructions,
        message=f"{instructions} instruction(s) appliquée(s) au schéma « {groupe.code} ».",
    )
